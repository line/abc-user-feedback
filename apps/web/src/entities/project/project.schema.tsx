/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { i18n } from 'next-i18next';
import { z } from 'zod';

export const projectSchema = z.object({
  id: z.number(),
  name: z
    .string()
    .min(1, { message: i18n?.t('hint.required') })
    .max(20, { message: i18n?.t('hint.max-length', { length: 20 }) }),
  description: z
    .string()
    .max(50, { message: i18n?.t('hint.max-length', { length: 50 }) })
    .nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  timezone: z.object({
    countryCode: z.string(),
    name: z.string(),
    offset: z.string(),
  }),
});

export const projectInfoSchema = projectSchema
  .pick({
    name: true,
    description: true,
    timezone: true,
  })
  .merge(z.object({ id: z.number().optional() }));
