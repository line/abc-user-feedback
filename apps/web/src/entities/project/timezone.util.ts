/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { getAllTimezones } from 'countries-and-timezones';

import type { Timezone } from './project.type';

export const getTimezoneOptions = () => {
  const result: Timezone[] = [];
  Object.values(getAllTimezones()).forEach((timezone) => {
    const { utcOffsetStr, countries, name } = timezone;
    countries.forEach((countryCode) => {
      result.push({
        countryCode,
        name,
        offset: utcOffsetStr,
      });
    });
  });
  return result;
};
export const getDefaultTimezone = () => {
  const options = getTimezoneOptions();
  const tz = options.find(
    (option) =>
      option.name === Intl.DateTimeFormat().resolvedOptions().timeZone,
  );
  if (!options[0]) {
    throw new Error("Can't find default timezone");
  }
  return tz ?? options[0];
};
