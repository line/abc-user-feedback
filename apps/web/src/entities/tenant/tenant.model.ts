/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import { create } from 'zustand';

import { client } from '@/shared';

import type { Tenant } from './tenant.type';

type State = { tenant: Tenant | null };

type Action = {
  setTenant: (tenant: Tenant) => void;
  refetchTenant: () => Promise<void>;
};

export const useTenantStore = create<State & Action>((set) => ({
  tenant: null,
  setTenant: (tenant) => set({ tenant }),
  refetchTenant: async () => {
    const { data } = await client.get({ path: '/api/admin/tenants' });
    set({ tenant: data });
  },
}));
