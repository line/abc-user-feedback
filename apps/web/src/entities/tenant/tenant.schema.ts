/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { z } from 'zod';

export const tenantSchema = z.object({
  id: z.number(),
  siteName: z.string(),
  description: z.string().nullable(),
  useEmail: z.boolean(),
  useOAuth: z.boolean(),
  isPrivate: z.boolean(),
  isRestrictDomain: z.boolean(),
  allowDomains: z
    .array(z.string().refine((v) => /[a-z]+\.[a-z]{2,3}/.test(v)))
    .nullable(),
  useEmailVerification: z.boolean(),
  oauthConfig: z
    .object({
      clientId: z.string(),
      emailKey: z.string(),
      scopeString: z.string(),
      clientSecret: z.string(),
      authCodeRequestURL: z.string(),
      accessTokenRequestURL: z.string(),
      userProfileRequestURL: z.string(),
    })
    .nullable(),
});

export const tenantInfoSchema = tenantSchema.pick({
  id: true,
  siteName: true,
  description: true,
});
export const authInfoScema = tenantSchema.pick({
  isPrivate: true,
  isRestrictDomain: true,
  allowDomains: true,
  useOAuth: true,
  useEmail: true,
  oauthConfig: true,
});
