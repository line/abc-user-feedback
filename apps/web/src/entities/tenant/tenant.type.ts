/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import { z } from 'zod';

const oauthConfigSchema = z.object({
  oauthUse: z.boolean(),
  clientId: z.string(),
  clientSecret: z.string(),
  authCodeRequestURL: z.string(),
  scopeString: z.string(),
  accessTokenRequestURL: z.string(),
  userProfileRequestURL: z.string(),
  emailKey: z.string(),
});

const tenantSchema = z.object({
  id: z.number(),
  siteName: z.string(),
  description: z.string().nullable(),
  useEmail: z.boolean(),
  useOAuth: z.boolean(),
  isPrivate: z.boolean(),
  isRestrictDomain: z.boolean(),
  allowDomains: z.array(z.string()),
  useEmailVerification: z.boolean(),
  oauthConfig: oauthConfigSchema.nullable(),
});

export type Tenant = z.infer<typeof tenantSchema>;

export type OAuthConfig = z.infer<typeof oauthConfigSchema>;
