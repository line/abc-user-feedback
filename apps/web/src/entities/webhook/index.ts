export * from './ui';
export * from './webhook.type';
