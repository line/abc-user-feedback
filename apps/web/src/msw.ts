/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';

import { signInWithOAuthMockHandlers } from './features/auth/sign-in-with-oauth/__mocks__/sign-in-with-oauth.mock-handler';
import type {
  OAIMethodPathKeys,
  OAIMethods,
  OAIPathParameters,
} from './shared';
import { convertToColonPath, getRequestUrl } from './shared';

export const server = setupServer(...signInWithOAuthMockHandlers);

export const simpleMockHttp = <
  TMethod extends OAIMethods,
  TPath extends OAIMethodPathKeys<TMethod>,
>({
  method,
  path,
  status = 200,
  params,
  data = {},
}: {
  method: TMethod;
  path: TPath;
  status?: 200 | 201 | 204 | 400 | 401 | 403 | 404 | 500;
  params?: OAIPathParameters<TPath, TMethod>;
  data?: Record<string, unknown>;
}) =>
  server.use(
    http[method](`${convertToColonPath(getRequestUrl(path, params))}`, () =>
      HttpResponse.json(data, { status }),
    ),
  );
