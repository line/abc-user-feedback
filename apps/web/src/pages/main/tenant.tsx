/**
 * Copyright 2025 LY Corporation
 *
 * LY Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import type { GetStaticProps } from 'next';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import { parseAsString, useQueryState } from 'nuqs';

import { Icon, Menu, MenuItem } from '@ufb/react';

import type { NextPageWithLayout } from '@/shared';
import SideMenuLayout from '@/shared/ui/side-menu-layout.ui';
import { Layout } from '@/widgets/layout';
import {
  LoginSetting,
  TenantInfoSetting,
  UserManagementSetting,
} from '@/widgets/setting-menu';

import serverSideTranslations from '@/server-side-translations';

const TenantPage: NextPageWithLayout = () => {
  const { t } = useTranslation();
  const router = useRouter();
  const [currentMenu] = useQueryState<string>(
    'menu',
    parseAsString.withDefault('tenant'),
  );

  return (
    <SideMenuLayout
      sideMenu={
        <Menu
          type="single"
          orientation="vertical"
          className="w-full p-0"
          value={currentMenu}
          onValueChange={(value) =>
            router.push({ pathname: router.pathname, query: { menu: value } })
          }
        >
          <MenuItem value="tenant">
            <Icon name="RiInformation2Line" />
            {t('tenant-setting-menu.tenant-info')}
          </MenuItem>
          <MenuItem value="login">
            <Icon name="RiUser2Line" />
            {t('tenant-setting-menu.sign-up-mgmt')}
          </MenuItem>
          <MenuItem value="user">
            <Icon name="RiShieldKeyholeLine" />
            {t('tenant-setting-menu.user-mgmt')}
          </MenuItem>
        </Menu>
      }
    >
      {currentMenu === 'tenant' && <TenantInfoSetting />}
      {currentMenu === 'login' && <LoginSetting />}
      {currentMenu === 'user' && <UserManagementSetting />}
    </SideMenuLayout>
  );
};

TenantPage.getLayout = (page) => {
  return <Layout title="Tenant">{page}</Layout>;
};

export const getStaticProps: GetStaticProps = async ({ locale }) => {
  return {
    props: {
      ...(await serverSideTranslations(locale)),
    },
  };
};

export default TenantPage;
