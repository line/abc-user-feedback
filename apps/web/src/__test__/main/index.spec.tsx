/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';

import { MOCK_PROJECTS } from '@/entities/project/__mocks__/project.mock-data';
import type { Tenant } from '@/entities/tenant';
import { useTenantStore } from '@/entities/tenant';

import { simpleMockHttp } from '@/msw';
import MainIndexPage from '@/pages/main';
import { render } from '@/test-utils';

describe('MainIndexPage', () => {
  test('should render without crashing', () => {
    useTenantStore.setState({ tenant: {} as Tenant });
    simpleMockHttp({
      method: 'get',
      path: '/api/admin/projects',
      status: 200,
      data: {
        meta: {
          itemCount: 1,
          totalItems: 1,
          itemsPerPage: 1,
          totalPages: 1,
          currentPage: 1,
        },
        items: MOCK_PROJECTS,
      },
    });
    for (const project of MOCK_PROJECTS) {
      simpleMockHttp({
        method: 'get',
        path: '/api/admin/projects/{projectId}/feedback-count',
        params: { projectId: project.id },
        status: 200,
        data: {
          total: faker.number.int(),
        },
      });
    }
    const page = MainIndexPage.getLayout?.(<MainIndexPage />);
    const { container } = render(<>{page}</>);
    expect(container).toMatchSnapshot();
  });
});
