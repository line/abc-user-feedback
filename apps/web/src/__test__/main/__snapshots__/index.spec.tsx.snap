// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainIndexPage should render without crashing 1`] = `
<div>
  <div
    class="min-w-[960px]"
  >
    <header
      class="relative flex h-[48px] items-center justify-between px-4"
    >
      <div
        class="flex flex-1 items-center gap-6"
      >
        <a
          class="flex cursor-pointer items-center gap-1"
          href="/auth/sign-in"
        >
          <img
            alt="logo"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/assets/images/logo.svg"
            style="color: transparent;"
            width="24"
          />
          <svg
            class="inline-block h-[24px] w-[123px]"
            height="24"
            width="24"
          />
        </a>
        <div
          class="flex items-center gap-2"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="bg-fill-tertiary inline-flex rounded-sm p-1.5"
            >
              <svg
                class="inline-block text-secondary"
                height="12"
                width="12"
              />
            </div>
            <span
              class="font-12-regular"
            />
          </div>
        </div>
      </div>
      <div
        class="flex items-center gap-2 self-stretch"
      >
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="btn btn-sm btn-secondary min-w-0 px-2"
            data-headlessui-state=""
            id="headlessui-listbox-button-:r2:"
            type="button"
          >
            <svg
              class="inline-block mr-1"
              height="16"
              width="16"
            />
            <span
              class="font-12-bold uppercase"
            />
          </button>
        </div>
        <button
          class="icon-btn icon-btn-sm icon-btn-secondary"
        >
          <svg
            class="inline-block text-primary"
            height="24"
            width="24"
          />
        </button>
      </div>
    </header>
    <div
      class="flex min-h-[calc(100vh-48px)] items-stretch"
    >
      <main
        class="mx-4 my-6 w-full overflow-x-auto"
      >
        <div
          class="mx-4 my-2 flex flex-col gap-8"
        >
          <section>
            <h1
              class="font-24-bold mb-4"
            >
              Tenant
            </h1>
            <div
              class="card h-[204px] w-[452px] rounded border p-8"
            >
              <div
                class="mb-10 flex gap-5"
              >
                <div
                  class="flex h-10 w-10 items-center justify-center rounded"
                  style="background-color: rgb(93, 123, 231);"
                >
                  <svg
                    class="inline-block text-inverse"
                    height="20"
                    width="20"
                  />
                </div>
                <div
                  class="flex-1"
                >
                  <p
                    class="font-16-bold mb-1"
                  />
                  <p
                    class="font-12-regular text-secondary line-clamp-1 break-all"
                  >
                    -
                  </p>
                </div>
              </div>
              <div
                class="flex gap-16"
              >
                <div>
                  <p
                    class="font-12-regular mb-1"
                  >
                    main.index.total-project
                  </p>
                  <p
                    class="font-24-bold"
                  >
                    0
                  </p>
                </div>
                <div>
                  <p
                    class="font-12-regular mb-1"
                  >
                    main.index.total-feedback
                  </p>
                  <p
                    class="font-24-bold"
                  >
                    0
                  </p>
                </div>
              </div>
            </div>
          </section>
          <section>
            <h1
              class="font-24-bold mb-4"
            >
              Project
            </h1>
            <div
              class="flex flex-wrap gap-4"
            >
              <div
                class="border-fill-tertiary flex h-[204px] w-[452px] flex-col items-center justify-center rounded border"
              >
                <button
                  class="btn btn-lg btn-primary w-[200px] gap-2"
                  data-state="closed"
                  disabled=""
                >
                  <svg
                    class="inline-block text-inverse"
                    height="24"
                    width="24"
                  />
                  main.index.create-project
                </button>
              </div>
            </div>
          </section>
        </div>
      </main>
    </div>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;
